app/services/search_orders.rb

```diff
@@ -19,7 +19,7 @@ def fetch_orders
       includes(:payments, :subscription, :shipments, :bill_address, :distributor, :order_cycle).
       ransack(params[:q])
 
-    @search.result(distinct: true).joins(:bill_address)
+    @search.result(distinct: true).left_joins(:bill_address)
```

Comment:
I was thinking that this join statement should be included earlier, then noticed that that the association is already mentioned in the `includes` statement. Wouldn't this have the effect of a left join, and so we can simply remove the  `joins(:bill_address)` part altogether?

I couldn't easily confirm this myself, because this bug wasn't covered by a spec.

Reply:
I decided to go ahead and inspect the SQL, current, and after removing `.left_joins(:bill_address)`( because bill_address is already included in `includes`):

```diff
 SELECT DISTINCT spree_addresses.*,
                 spree_orders.*
 FROM   spree_orders
        LEFT OUTER JOIN spree_line_items
                     ON spree_line_items.order_id = spree_orders.id
        LEFT OUTER JOIN spree_addresses
                     ON spree_addresses.id = spree_orders.bill_address_id
-       LEFT OUTER JOIN spree_addresses bill_addresses_spree_orders
-                    ON bill_addresses_spree_orders.id = spree_orders.bill_address_id
 WHERE  ( spree_orders.distributor_id IN (SELECT enterprises.id
                                          FROM   enterprises)
           OR spree_orders.order_cycle_id IN (SELECT order_cycles.id
                                              FROM   order_cycles) )
        AND ( spree_line_items.id IS NOT NULL
               OR spree_orders.state IN ( 'complete', 'canceled', 'resumed',
                                          'awaiting_return', 'returned' ) )
 ORDER  BY spree_addresses.lastname ASC
```
(I used an online code formatter for legibility too)

We can see that there was a duplicate LEFT OUTER JOIN on `spree_addresses`, which doesn't seem to be referred to at all. I tested this with the new specs and it continues to pass.

So I think we can remove it, do others agree?

Reply:
Actually I have to admit I don't understand why `bill_adresses` is included with a left join, while other tables in the `includes` aren't.
In any case, the tests seem to indicate that we don't need the additional bill_addresses join, so I've pushed a suggested commit to demonstrate. Feel free to remove again if it's inappropriate.

Reply:
> Actually I have to admit I don't understand why bill_adresses is included with a left join, while other tables in the includes aren't.

The Rails `includes` method can use either a left join or a separate query:

> With includes, Active Record ensures that all of the specified associations are loaded using the minimum possible number of queries.
> https://guides.rubyonrails.org/active_record_querying.html#eager-loading-associations

There are two ways to tell Rails more specifically what you want:

* `references`: Tell Rails to use a left join because you have `select` or `where` statements referencing that table.
* `eager_load`: Instead of `includes`, this one always does a left join. It's like `includes(:table).references(:table)`.
* `preload`: Instead of `includes`, this one always loads in a separate query.


So Rails has to basic mechanisms: eager_load and preload. If you don't care which one is used and you want Rails to decide that when the full query is executed then you can use includes and it will use one of them under the hood. If you want to automatic `includes` in some cases but in some you want to reference the the table then you should declare that.

For this particular case of bill_address, Rails seems to do the right thing already. I'm wondering if Ransack is clever enough to tell Rails that it's referencing that association. Or maybe we do something already that makes Rails think that we need that join. Since we do a `select` we should probably add the `references` to be sure Rails joins. But since there were two joins before, Rails may see two different associations. We need to check that `references` isn't adding one again. Is it maybe that we use `bill_address` in some cases and `billing_address` in others? I noticed somewhere that we using *billing* in a Ransack query because there were issues with *bill* but maybe with this knowledge there's a better way to resolve it.

Reply:
Wow, it's complex. Thanks for diving in, but I'm not feeling super clear if it's worth changing anything?

It appears we have it covered by specs, and they are passing so I'd avoid adding new code.
But... I just tried adding a spec for _billing_ address, and it broke due to my change. Ok I will revert my change!

But... the order is wrong for _billing_ address, so we don't seem to support a query on this alias. Do we need to?